src/Integration.Vsix/Helpers/ActiveDocumentLocator.cs

```diff
@@ -0,0 +1,78 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.TextManager.Interop;
+
+namespace SonarLint.VisualStudio.Integration.Vsix
+{
+    internal interface IActiveDocumentLocator
+    {
+        /// <summary>
+        /// Returns the current active document, or null if there is no active document
+        /// </summary>
+        /// <returns></returns>
+        ITextDocument FindActiveDocument();
+    }
+
+    internal class ActiveDocumentLocator : IActiveDocumentLocator
```

Comment:
New class to handle the ugly part of finding the active VS document and extracting the relevant text document. The required steps are:
* using the monitor selection service to the current doc frame, which should be an `IVsWindowFrame`
* get the backing DocData (essentially the data storage for the frame - multiple frames can have the same DocData). It should be an `IVsTextBuffer`.
* the `IVsTextBuffer` is a legacy VS interface. We need to use the adapter service to convert it to the new-style interface. See https://docs.microsoft.com/en-us/visualstudio/extensibility/legacy-interfaces-in-the-editor?view=vs-2015.
* the `ITextBuffer` should have an associated `ITextDocument`, which is what we want to return.